File: CodeGenDotNet 159059-1372 (a)/CSharp/Chapter 3/BruteForceExample/BruteForceExample.cs
using System;

namespace KADGen.BruteForceExample
{
	/// <summary>
	/// Example file outputting the SimpleDataContainer class
	/// </summary>
	public class BruteForceExample
	{
		public static System.IO.Stream GetStream(	string name,
													string fileName,
													string genDateTime,
													System.Xml.XmlNode nodeSelect)
		{
			System.Console.WriteLine("Test");
			System.Console.ReadLine();
			System.IO.MemoryStream stream = new System.IO.MemoryStream();
			System.CodeDom.Compiler.IndentedTextWriter inWriter = new System.CodeDom.Compiler.IndentedTextWriter( new System.IO.StreamWriter( stream ) );

			System.Xml.XmlNodeList nodeList;
			string singularName = Utility.Tools.GetAttributeOrEmpty( nodeSelect, "SingularName" );
			System.Xml.XmlNamespaceManager nsmgr = Utility.Tools.BuildNamespaceManager( nodeSelect.OwnerDocument, "dbs", false );

			Support.FileOpen( inWriter, "KADGen,System", fileName, genDateTime);
			inWriter.WriteLine( "" );
			Support.WriteLineAndIndent( inWriter, "public class " + singularName + "Collection : CollectionBase" );
			CollectionConstructors( inWriter, nsmgr, nodeSelect );
			PublicAndFriend( inWriter, nsmgr, nodeSelect );
			Support.WriteLineAndOutdent( inWriter, "}" );

			inWriter.WriteLine( "" );
			inWriter.WriteLine( "" );
			Support.WriteLineAndIndent( inWriter, "public class " + singularName + " : RowBase" );
			ClassLevelDeclarations( inWriter, nsmgr, nodeSelect);
			Constructors( inWriter, nsmgr, nodeSelect );
			BaseClassImplementation( inWriter, nsmgr, nodeSelect );
			Support.MakeRegion( inWriter, "Field access properties" );
			nodeList = nodeSelect.SelectNodes( "dbs:TableColumns/*", nsmgr );
			foreach( System.Xml.XmlNode nodeColumn in nodeList )
			{
				ColumnMethods( inWriter, nsmgr, nodeColumn );
			}
			Support.EndRegion( inWriter );
			Support.WriteLineAndIndent( inWriter, "End Class" );

			inWriter.Flush();
			return stream;
		}

		public static void CollectionConstructors(	System.CodeDom.Compiler.IndentedTextWriter inWriter,
													System.Xml.XmlNamespaceManager nsmgr,
													System.Xml.XmlNode node )
		{
			Support.MakeRegion( inWriter, "Constructors" );
			string className = Utility.Tools.GetAttributeOrEmpty(node, "SingularName") + "Collection";
			inWriter.WriteLine( "protected " + className + "() : base( \"" + className + "\" )" );
			Support.WriteLineAndIndent( inWriter, "{" );
			Support.WriteLineAndOutdent( inWriter, "}" );
			Support.EndRegion( inWriter );
		}

		public static void PublicAndFriend(	System.CodeDom.Compiler.IndentedTextWriter inWriter,
											System.Xml.XmlNamespaceManager nsmgr,
											System.Xml.XmlNode node )
		{
			Support.MakeRegion( inWriter, "Public and Friend Properties, Methods and Events" );
			System.Xml.XmlNode nodeTemp = node.SelectSingleNode( "dbs:TableConstraints/dbs:PrimaryKey/dbs:PKField", nsmgr );
			string primaryKeyName = "";
			System.Xml.XmlNode primaryKey;
			if( nodeTemp != null )
			{
				primaryKeyName = Utility.Tools.GetAttributeOrEmpty( nodeTemp, "Name" );
			}
			primaryKey = node.SelectSingleNode( "dbs:TableColumns/dbs:TableColumn[@Name='" + primaryKeyName + "']", nsmgr );
			inWriter.WriteLine( "public overloads void Fill( " );
			inWriter.Indent += 4;
			if( primaryKey != null )
			{
				inWriter.WriteLine( Utility.Tools.GetAttributeOrEmpty( primaryKey, "NETType" ) + primaryKeyName + "," );
			}
			inWriter.WriteLine( "int UserID" );
			inWriter.Indent -= 4;
			Support.WriteLineAndIndent( inWriter, "{" );
			inWriter.Write( Utility.Tools.GetAttributeOrEmpty( node, "SingularName" ) + "DataAccessor.Fill( this, " );
			if( primaryKey != null )
			{
				inWriter.Write( primaryKeyName + ", " );
			}
			inWriter.WriteLine( "UserID );" );
			Support.WriteLineAndOutdent( inWriter, "}" );
			Support.EndRegion( inWriter );
		}

		public static void ClassLevelDeclarations(	System.CodeDom.Compiler.IndentedTextWriter inWriter,
													System.Xml.XmlNamespaceManager nsmgr,
													System.Xml.XmlNode node )
		{
			System.Xml.XmlNodeList nodeList;
			string pre = "";
			Support.MakeRegion( inWriter, "Class Level Declarations" );
			inWriter.WriteLine( string.Format( "protected {0} mCollection;", Utility.Tools.GetAttributeOrEmpty( node, "SingularName" ) ) );
			inWriter.WriteLine( "private static int mNextPrimaryKey = -1;" );
			nodeList = node.SelectNodes( "dbs:TableColumns/*", nsmgr );
			inWriter.WriteLine( "" );
			foreach( System.Xml.XmlNode nodeSub in nodeList )
			{
				if( Utility.Tools.GetAttributeOrEmpty(nodeSub, "NETType").Length == 0 )
				{
					pre = "// ";
				}
				inWriter.WriteLine( string.Format( "{0}private {1} {2};", pre, Utility.Tools.GetAttributeOrEmpty(nodeSub, "NETType"), Utility.Tools.GetAttributeOrEmpty(nodeSub, "Name") ) );
			}
			Support.EndRegion( inWriter );
		}
		private static void Constructors(	System.CodeDom.Compiler.IndentedTextWriter inWriter,
											System.Xml.XmlNamespaceManager nsmgr,
											System.Xml.XmlNode node )
		{
			string singularName = Utility.Tools.GetAttributeOrEmpty( node, "SingularName" );
			Support.MakeRegion( inWriter, "Constructors" );
			inWriter.WriteLine( "internal {0}( {1} {1} ) : base()", singularName, singularName+"Collection" );
			Support.WriteLineAndIndent( inWriter, "{" );
			inWriter.WriteLine( "mCollection = " + singularName + "Collection" );
			Support.WriteLineAndOutdent( inWriter, "}" );
			Support.EndRegion( inWriter );
		}
		private static void BaseClassImplementation(	System.CodeDom.Compiler.IndentedTextWriter inWriter,
													System.Xml.XmlNamespaceManager nsmgr,
													System.Xml.XmlNode node )
		{
			Support.MakeRegion( inWriter, "Base Class Implementation" );
			System.Xml.XmlNode nodeTemp = node.SelectSingleNode( "dbs:TableConstraints/dbs:PrimaryKey/dbs:PKField", nsmgr );
			string primaryKeyName = "";
			System.Xml.XmlNode primaryKey;
			if( nodeTemp != null )
			{
				primaryKeyName = Utility.Tools.GetAttributeOrEmpty( nodeTemp, "Name" );
			}
			primaryKey = node.SelectSingleNode( "dbs:TableColumns/dbs:TableColumn[@Name='" + primaryKeyName + "']", nsmgr );
			if( primaryKey!= null )
			{
				if( Utility.Tools.GetAttributeOrEmpty( primaryKey, "IsAutoIncrement" ) == "1" )
				{
					inWriter.WriteLine( "internal void SetNewPrimaryKey()" );
					Support.WriteLineAndIndent( inWriter, "{" );
					inWriter.WriteLine( primaryKeyName + " = mNextPrimaryKey" );
					inWriter.WriteLine( "mNextPrimaryKey --" );
					Support.WriteLineAndOutdent( inWriter, "}" );
				}
			}
			Support.EndRegion( inWriter );
		}
		private static void ColumnMethods(	System.CodeDom.Compiler.IndentedTextWriter inWriter,
			System.Xml.XmlNamespaceManager nsmgr,
			System.Xml.XmlNode node )
		{
			string netTypeName = Utility.Tools.GetAttributeOrEmpty( node, "NETType" );
			string name = Utility.Tools.GetAttributeOrEmpty( node, "Name" );
			if( netTypeName.Trim().Length == 0 )
			{
				inWriter.WriteLine( "" );
				inWriter.WriteLine( "// Column " + name + " is not included because it uses" );
				inWriter.WriteLine( "// a SQLType (" + netTypeName + ") that is not yet supported" );
				inWriter.WriteLine( "" );
			}
			else
			{
				inWriter.WriteLine( "public ColumnInfo " + name + "()" );
				Support.WriteLineAndIndent( inWriter, "{" );
				inWriter.WriteLine( "ColumnInfo columnInfo = new ColumnInfo" );
				inWriter.WriteLine( "columnInfo.FieldName = " + Support.DQ + name + Support.DQ );
				inWriter.WriteLine( "columnInfo.FieldType = GetType(" + netTypeName + ")" );
				inWriter.WriteLine( "columnInfo.SQLType = " + Support.DQ + Utility.Tools.GetAttributeOrEmpty( node, "SQLType" ) + Support.DQ );
				inWriter.WriteLine( "columnInfo.Caption = " + Support.DQ + Utility.Tools.GetAttributeOrEmpty( node, "Caption" ) + Support.DQ );
				inWriter.WriteLine( "columnInfo.Desc = " + Support.DQ + Utility.Tools.GetAttributeOrEmpty( node, "Desc" ) + Support.DQ );
				inWriter.WriteLine( "return columnInfo" );
				Support.WriteLineAndOutdent( inWriter, "}" );

				inWriter.WriteLine( "" );
				inWriter.WriteLine( "public " + netTypeName + " " + name );
				Support.WriteLineAndIndent( inWriter, "{" );
				inWriter.WriteLine( "get" );
				Support.WriteLineAndIndent( inWriter, "{" );
				inWriter.WriteLine( "return m" + name + ";" );
				Support.WriteLineAndOutdent(inWriter, "}");
				inWriter.WriteLine( "set" );
				Support.WriteLineAndIndent( inWriter, "{" );
				inWriter.WriteLine( "m" + name + " = value;" );
				Support.WriteLineAndOutdent( inWriter, "}" );
				Support.WriteLineAndOutdent( inWriter, "}" );
			}

		}
	}
}

File: CodeGenDotNet 159059-1372 (a)/VBCode/Chapter 3/BruteForceExample/BruteForceExample.vb
Option Strict On
Option Explicit On 

Imports System

#Region "Description"
'Example file outputting the SimpleDataContainer class
#End Region

Public Class BruteForceExample
   Public Shared Function GetStream( _
                  ByVal Name As String, _
                  ByVal filename As String, _
                  ByVal gendatetime As String, _
                  ByVal nodeSelect As Xml.XmlNode) _
                  As IO.Stream
      Dim stream As New IO.MemoryStream
      Dim inwriter As New CodeDom.Compiler.IndentedTextWriter( _
                  New IO.StreamWriter(stream))
      Dim nodeList As Xml.XmlNodeList
      Dim nodeColumn As Xml.XmlNode
      Dim singularName As String = Utility.Tools.GetAttributeOrEmpty(nodeSelect, _
                  "SingularName")
      Dim nsmgr As Xml.XmlNamespaceManager = _
                  Utility.Tools.BuildNameSpaceManager( _
                        nodeSelect.OwnerDocument, "dbs", False)

      System.Console.WriteLine("Test VB")
      System.Console.ReadLine()
      Support.FileOpen(inwriter, "KADGen,System", filename, gendatetime)
      inwriter.WriteLine("")
      Support.WriteLineAndIndent(inwriter, "Public Class " & singularName & _
                  "Collection")
      inwriter.WriteLine("Inherits CollectionBase")
      CollectionConstructors(inwriter, nsmgr, nodeSelect)
      PublicAndFriend(inwriter, nsmgr, nodeSelect)
      Support.WriteLineAndOutdent(inwriter, "End Class")

      inwriter.WriteLine("")
      inwriter.WriteLine("")
      Support.WriteLineAndIndent(inwriter, "Public Class " & singularName)
      inwriter.WriteLine("Inherits RowBase")
      ClassLevelDeclarations(inwriter, nsmgr, nodeSelect)
      Constructors(inwriter, nsmgr, nodeSelect)
      BaseClassImplementation(inwriter, nsmgr, nodeSelect)
      Support.MakeRegion(inwriter, "Field access properties")
      nodeList = nodeSelect.SelectNodes("dbs:TableColumns/*", nsmgr)
      For Each nodeColumn In nodeList
         ColumnMethods(inwriter, nsmgr, nodeColumn)
      Next
      Support.EndRegion(inwriter)
      Support.WriteLineAndIndent(inwriter, "End Class")

      inwriter.Flush()
      Return stream
   End Function


   Private Shared Sub CollectionConstructors( _
                  ByVal inWriter As CodeDom.Compiler.IndentedTextWriter, _
                  ByVal nsmgr As Xml.XmlNamespaceManager, _
                  ByVal node As Xml.XmlNode)
      Support.MakeRegion(inWriter, "Constructors")
      Support.WriteLineAndIndent(inWriter, "Protected Sub New()")
      inWriter.WriteLine("MyBase.New(" & Support.DQ & _
                     Utility.Tools.GetAttributeOrEmpty(node, "SingularName") & _
                     "Collection" & Support.DQ & ")")
      Support.WriteLineAndOutdent(inWriter, "End Sub")
      Support.EndRegion(inWriter)
   End Sub


   Private Shared Sub PublicAndFriend( _
                  ByVal inWriter As CodeDom.Compiler.IndentedTextWriter, _
                  ByVal nsmgr As Xml.XmlNamespaceManager, _
                  ByVal node As Xml.XmlNode)
      Support.MakeRegion(inWriter, _
                  "Public and Friend Properties, Methods and Events")
      Dim nodeTemp As Xml.XmlNode = node.SelectSingleNode( _
                  "dbs:TableConstraints/dbs:PrimaryKey/dbs:PKField", nsmgr)
      Dim primaryKeyName As String = ""
      Dim primaryKey As Xml.XmlNode
      If Not nodeTemp Is Nothing Then
         primaryKeyName = Utility.Tools.GetAttributeOrEmpty(nodeTemp, "Name")
      End If
      primaryKey = node.SelectSingleNode( _
                  "dbs:TableColumns/dbs:TableColumn[@Name='" & primaryKeyName & _
                  "']", nsmgr)
      Support.WriteLineAndIndent(inWriter, "Public Overloads Sub Fill( _")
      inWriter.Indent += 4
      If Not primaryKey Is Nothing Then
         inWriter.Write("ByVal " & primaryKeyName & " As ")
         inWriter.WriteLine(Utility.Tools.GetAttributeOrEmpty(primaryKey, _
                  "NETType") & ", _")
      End If
      inWriter.WriteLine("ByVal UserID As Int32)")
      inWriter.Indent -= 4
      inWriter.WriteLine("ByVal UserID As Int32)")
      inWriter.Write(Utility.Tools.GetAttributeOrEmpty(node, "SingularName") & _
                    "DataAccessor.Fill(Me")
      If Not primaryKey Is Nothing Then
         inWriter.Write(", " & primaryKeyName)
      End If
      inWriter.WriteLine(", UserID)")
      Support.WriteLineAndOutdent(inWriter, "End Sub")
      Support.EndRegion(inWriter)
   End Sub

   Private Shared Sub ClassLevelDeclarations( _
                  ByVal inWriter As CodeDom.Compiler.IndentedTextWriter, _
                  ByVal nsmgr As Xml.XmlNamespaceManager, _
                  ByVal node As Xml.XmlNode)
      Dim nodeList As Xml.XmlNodeList
      Dim pre As String = ""
      Support.MakeRegion(inWriter, "Class Level Declarations")
      inWriter.WriteLine("Protected mCollection As " & _
                  Utility.Tools.GetAttributeOrEmpty(node, "SingularName") & _
                  "Collection")
      inWriter.WriteLine("Private Shared mNextPrimaryKey As Int32 = -1")
      nodeList = node.SelectNodes("dbs:TableColumns/*", nsmgr)
      inWriter.WriteLine("")
      For Each nodeSub As Xml.XmlNode In nodeList
         If Utility.Tools.GetAttributeOrEmpty(nodeSub, "NETType").Length = 0 Then
            pre = "' "
         End If
         inWriter.WriteLine(pre & "Private m" & _
                  Utility.Tools.GetAttributeOrEmpty(nodeSub, "Name") & " As " & _
                  Utility.Tools.GetAttributeOrEmpty(nodeSub, "NETType"))
      Next
      Support.EndRegion(inWriter)
   End Sub

   Private Shared Sub Constructors( _
                  ByVal inWriter As CodeDom.Compiler.IndentedTextWriter, _
                  ByVal nsmgr As Xml.XmlNamespaceManager, _
                  ByVal node As Xml.XmlNode)
      Dim singularName As String = _
                  Utility.Tools.GetAttributeOrEmpty(node, "SingularName")
      Support.MakeRegion(inWriter, "Constructors")
      Support.WriteLineAndIndent(inWriter, "Friend Sub New(ByVal " & _
                  singularName & "Collection As " & singularName & "Collection)")
      inWriter.WriteLine("MyBase.new()")
      inWriter.WriteLine("mCollection = " & singularName & "Collection")
      Support.WriteLineAndOutdent(inWriter, "End Sub")
      Support.EndRegion(inWriter)

   End Sub

   Private Shared Sub BaseClassImplementation( _
                  ByVal inWriter As CodeDom.Compiler.IndentedTextWriter, _
                  ByVal nsmgr As Xml.XmlNamespaceManager, _
                  ByVal node As Xml.XmlNode)
      Support.MakeRegion(inWriter, "Base Class Implementation")
      Dim nodeTemp As Xml.XmlNode = node.SelectSingleNode( _
                  "dbs:TableConstraints/dbs:PrimaryKey/dbs:PKField", nsmgr)
      Dim primaryKeyName As String = ""
      Dim primaryKey As Xml.XmlNode
      If Not nodeTemp Is Nothing Then
         primaryKeyName = Utility.Tools.GetAttributeOrEmpty(nodeTemp, "Name")
      End If
      primaryKey = node.SelectSingleNode( _
               "dbs:TableColumns/dbs:TableColumn[@Name='" & primaryKeyName & "']", _
               nsmgr)
      If Not primaryKey Is Nothing Then
         If Utility.Tools.GetAttributeOrEmpty(primaryKey, "IsAutoIncrement") = "1" _
                  Then
            Support.WriteLineAndIndent(inWriter, "Friend Sub SetNewPrimaryKey()")
            inWriter.WriteLine(primaryKeyName & " = mNextPrimaryKey")
            inWriter.WriteLine("mNextPrimaryKey -= 1")
            Support.WriteLineAndOutdent(inWriter, "End Sub")
         End If
      End If
      Support.EndRegion(inWriter)

   End Sub


   Private Shared Sub ColumnMethods( _
                  ByVal inWriter As CodeDom.Compiler.IndentedTextWriter, _
                  ByVal nsmgr As Xml.XmlNamespaceManager, _
                  ByVal node As Xml.XmlNode)
      Dim netTypeName As String = Utility.Tools.GetAttributeOrEmpty(node, _
                  "NETType")
      Dim name As String = Utility.Tools.GetAttributeOrEmpty(node, "Name")
      If netTypeName.Trim.Length = 0 Then
         inWriter.WriteLine("")
         inWriter.WriteLine("' Column " & name & _
                  " is not included because it uses")
         inWriter.WriteLine("' a SQLType (" & netTypeName & _
                           ") that is not yet supported")
         inWriter.WriteLine("")
      Else
         Support.WriteLineAndIndent(inWriter, "Public Function " & name & _
                  "ColumnInfo As ColumnInfo")
         inWriter.WriteLine("Dim columnInfo As New ColumnInfo")
         inWriter.WriteLine("columnInfo.FieldName = " & Support.DQ & name & _
                  Support.DQ)
         inWriter.WriteLine("columnInfo.FieldType = GetType(" & netTypeName & ")")
         inWriter.WriteLine("columnInfo.SQLType = " & Support.DQ & _
                  Utility.Tools.GetAttributeOrEmpty(node, "SQLType") & Support.DQ)
         inWriter.WriteLine("columnInfo.Caption = " & _
                  Support.DQ & _
                  Utility.Tools.GetAttributeOrEmpty(node, "Caption") & Support.DQ)
         inWriter.WriteLine("columnInfo.Desc = " & _
                  Support.DQ & Utility.Tools.GetAttributeOrEmpty(node, "Desc") & _
                  Support.DQ)
         inWriter.WriteLine("Return columnInfo")
         Support.WriteLineAndOutdent(inWriter, "End Function")

         inWriter.WriteLine("")
         Support.WriteLineAndIndent(inWriter, "Public Property " & name & _
                  " As " & netTypeName)
         Support.WriteLineAndIndent(inWriter, "Get")
         inWriter.WriteLine("Return m" & name)
         Support.WriteLineAndOutdent(inWriter, "End Get")
         Support.WriteLineAndIndent(inWriter, "Set(ByVal Value As " & _
                  netTypeName & ")")
         inWriter.WriteLine("m" & name & " = Value")
         Support.WriteLineAndOutdent(inWriter, "End Set")
         Support.WriteLineAndOutdent(inWriter, "End Property")
      End If

   End Sub
End Class

